WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,243 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using WalletWasabi.Exceptions;
+using WalletWasabi.Helpers;
+using WalletWasabi.Interfaces.EventSourcing;
+
+namespace WalletWasabi.EventSourcing
+{
+	/// <summary>
+	/// Thread safe without locks in memory event repository implementation
+	/// </summary>
+	public class InMemoryEventRepository : IEventRepository
+	{
+		private static readonly IReadOnlyList<WrappedEvent> EmptyResult
+			= ImmutableList<WrappedEvent>.Empty;
+
+		private static readonly IReadOnlyList<string> EmptyIds
+			= ImmutableList<string>.Empty;
+
+		private static readonly IComparer<WrappedEvent> WrappedEventSequenceIdComparer
+			= Comparer<WrappedEvent>.Create((a, b) => a.SequenceId.CompareTo(b.SequenceId));
+
+		private ConcurrentDictionary
+			// aggregateType
+			<string,
+			ConcurrentDictionary
+				// aggregateId
+				<string,
+				(
```

Comment:
I would avoid tuples as much as possible. Mentioned here https://github.com/zkSNACKs/WalletWasabi/pull/6724#discussion_r760837032.

Reply:
it was accepted by @molnard

Reply:
I remember that I wrote somewhere to use a record. Anyway if not then my bad 🙏 ,  please use a record instead of tuple when possible.

Reply:
I was sitting next to you in the office you told it to my face. You said: "This is unusual but I like it is readable."

Reply:
And this is not about right or wrong. This is subjective opinion.

Reply:
@molnard

> Why I cannot see the mentioned comment?

That's because Github has some issues once PR has a lot of review comments. This is the one, I believe:

![image](https://user-images.githubusercontent.com/58662979/145038518-eca16df4-ddcd-4886-9fb4-469452457a16.png)

@bitcoin-eagle 

> And this is not about right or wrong. This is subjective opinion.

It is not about being right or wrong. 

At this point I'm not even sure if you have contributed to an open-source project, because (at least in my experience) almost every project has strict rules regarding formatting and if somebody wants to contribute he/she has to respect that. The rules are mostly documented only partially (because it would be probably long and nobody would read it anyway) and people simply try to imitate the code style they see in other files. I agree it would be great to have it automated though.

Then second point is that, yes, it is subjective whether tuples are more readable or structs. But the key point here is that somebody will maintain that code, others will read it, so you don't write the code for yourself but for others. But also the review comments are not to be taken personally like "it was a mistake to use tuples", it's simply that people prefer some some approaches based on their experience what worked for them and what didn't - this might be hard to accept but that's a price for collaborating with people.

Reply:
I didn't claim to you or anyone else that "I have contributed to an open-source project." You unsubstantiated expectations are not really my responsibility, are they?

Reply:
"strict rules" and "documented only partially" is good basis for gasslighting, toxic emotional manipulation and abuses of power. That just means that somebody can do anything and change the "implicit rules" and somebody else needs to listen no matter what.

There is a good reason why rules needs to be well defined and in a written form.

Reply:
And it is unacceptable for me to follow "strict rules" which are "documented only partially". If you want to follow coding guidelines and block my pull requests because of that please update and point me to a documentation everytime that happens 🙏

Reply:
you can start with the white-space around ifs if that is so important to the project